pulp_container/app/registry_api.py

```diff
@@ -684,54 +705,57 @@ def put(self, request, path, pk=None):
 
-        with NamedTemporaryFile("ab") as temp_file:
-            for chunk in chunks:
-                temp_file.write(chunk.file.read())
-                chunk.file.close()
-            temp_file.flush()
+        if chunks.exists():
 
-            uploaded_file = PulpTemporaryUploadedFile.from_file(File(open(temp_file.name, "rb")))
+            with NamedTemporaryFile("ab") as temp_file:
+                for chunk in chunks:
+                    temp_file.write(chunk.file.read())
+                    chunk.file.close()
+                temp_file.flush()
 
-        if uploaded_file.hashers["sha256"].hexdigest() == digest[len("sha256:") :]:
+                uploaded_file = PulpTemporaryUploadedFile.from_file(
+                    File(open(temp_file.name, "rb"))
+                )
+            if uploaded_file.hashers["sha256"].hexdigest() != digest[len("sha256:") :]:
+                raise Exception("The digest did not match")
             try:
                 artifact = Artifact.init_and_validate(uploaded_file)
                 artifact.save()
             except IntegrityError:
                 artifact = Artifact.objects.get(sha256=artifact.sha256)
                 artifact.touch()
-            try:
-                blob = models.Blob(digest=digest)
-                blob.save()
-            except IntegrityError:
-                blob = models.Blob.objects.get(digest=digest)
-                blob.touch()
-            try:
-                blob_artifact = ContentArtifact(
-                    artifact=artifact, content=blob, relative_path=digest
-                )
-                blob_artifact.save()
-            except IntegrityError:
-                ca = ContentArtifact.objects.get(content=blob, relative_path=digest)
-                if not ca.artifact:
-                    ca.artifact = artifact
-                    ca.save(update_fields=["artifact"])
-            upload.delete()
-
-            dispatched_task = dispatch(
-                add_and_remove,
-                shared_resources=[f"upload:{pk}"],
-                exclusive_resources=[repository],
-                kwargs={
-                    "repository_pk": str(repository.pk),
-                    "add_content_units": [str(blob.pk)],
-                    "remove_content_units": [],
-                },
-            )
-
-            if has_task_completed(dispatched_task):
-                return BlobResponse(blob, path, 201, request)
         else:
-            raise Exception("The digest did not match")
+            artifact = Artifact.objects.get(sha256=digest[len("sha256:") :])
+            artifact.touch()
+        try:
+            blob = models.Blob(digest=digest)
+            blob.save()
+        except IntegrityError:
+            blob = models.Blob.objects.get(digest=digest)
+            blob.touch()
+        try:
+            blob_artifact = ContentArtifact(artifact=artifact, content=blob, relative_path=digest)
+            blob_artifact.save()
```

Comment:
The Blob and it's content-artifact should be created in a transaction. So the Blob once released into existence will already have it's artifact attached.